src/common/hist_util.cc

```diff
@@ -707,43 +707,123 @@ void SubtractionHist(GHistRow dst, const GHistRow src1, const GHistRow src2,
+        PREFETCH_READ_T0(index + j);
+      }
+    }
+
+    for (size_t j = icol_start; j < icol_start + n_features; ++j) {
+      const uint32_t idx_bin = 2 * index[j];
+
+      hist_data[idx_bin]   += pgh[idx_gh];
+      hist_data[idx_bin+1] += pgh[idx_gh+1];
+    }
+  }
+}
+
+template<typename FPType, bool do_prefetch>
+void BuildHistSparseKernel(const size_t* rid, const float* pgh, const uint32_t* gradient_index,
+                           FPType* hist_data, const size_t* row_ptr, size_t ibegin, size_t iend) {
+  for (size_t i = ibegin; i < iend; ++i) {
     const size_t icol_start = row_ptr[rid[i]];
     const size_t icol_end = row_ptr[rid[i]+1];
+    const size_t idx_gh = 2*rid[i];
+
+    if (do_prefetch) {
+      const size_t icol_start_prftch = row_ptr[rid[i+Prefetch::kPrefetchOffset]];
+      const size_t icol_end_prefect = row_ptr[rid[i+Prefetch::kPrefetchOffset]+1];
 
-    if (i < nrows - no_prefetch_size) {
-      PREFETCH_READ_T0(row_ptr + rid[i + prefetch_offset]);
-      PREFETCH_READ_T0(pgh + 2*rid[i + prefetch_offset]);
+      PREFETCH_READ_T0(pgh + 2*rid[i + Prefetch::kPrefetchOffset]);
+      for (size_t j = icol_start_prftch; j < icol_end_prefect; j+=Prefetch::kPrefetchStep) {
+        PREFETCH_READ_T0(gradient_index + j);
+      }
     }
 
     for (size_t j = icol_start; j < icol_end; ++j) {
-      const uint32_t idx_bin = 2*index[j];
-      const size_t idx_gh = 2*rid[i];
-
-      hist_data[idx_bin] += pgh[idx_gh];
+      const uint32_t idx_bin = 2 * gradient_index[j];
+      hist_data[idx_bin]   += pgh[idx_gh];
       hist_data[idx_bin+1] += pgh[idx_gh+1];
     }
   }
 }
 
+template<typename FPType, bool do_prefetch>
+void BuildHistKernel(const size_t* rid, const float* pgh, const uint32_t* gradient_index,
+                     FPType* hist_data, const size_t* row_ptr, size_t ibegin, size_t iend,
+                     bool isDense) {
+  if (isDense) {
+    const size_t n_features = row_ptr[rid[0]+1] - row_ptr[rid[0]];
+    BuildHistDenseKernel<FPType, do_prefetch>(rid, pgh, gradient_index, hist_data,
+                                              ibegin, iend, n_features);
+  } else {
+    BuildHistSparseKernel<FPType, do_prefetch>(rid, pgh, gradient_index, hist_data, row_ptr,
+                                               ibegin, iend);
+  }
+}
+
+void GHistBuilder::BuildHist(const std::vector<GradientPair>& gpair,
+                             const RowSetCollection::Elem row_indices,
+                             const GHistIndexMatrix& gmat,
+                             GHistRow hist,
+                             bool isDense) {
+  const size_t* rid =  row_indices.begin;
+  const size_t nrows = row_indices.Size();
+  const uint32_t* index = gmat.index.data();
+  const size_t* row_ptr =  gmat.row_ptr.data();
+  const float* pgh = reinterpret_cast<const float*>(gpair.data());
+
+  using FPType = decltype(tree::GradStats::sum_grad);
+  FPType* hist_data = reinterpret_cast<FPType*>(hist.data());
```

Comment:
I added this to handle any FP types in GradStats without code changes in other places except GradStats itself:
```cpp
using FPType = decltype(tree::GradStats::sum_grad);
```
Now it's double, but it looks like that it can be changed to float without accuracy loosing.